### PR TITLE
Added testcase codename validator

### PR DESCRIPTION
**Pull request by akmohtashami**
For details, please refer to #582.

---

This change is [Review on Reviewable](https://reviewable.io/reviews/cms-dev/cms/584)

### CODE REVIEW

File: cms/db/task.py [hunk 1]
```diff
@@ -555,3 +557,8 @@ class Testcase(Base):
     output = Column(
         String,
         nullable=False)
+
+    @validates("codename")
+    def validate_codename(self, key, codename):
```
Comment (stefano-maggiolo) [inline cms/db/task.py]: Codacy Issue found: [Method could be a function](https://www.codacy.com/app/s-maggiolo/cms/file/2228608531/issues/source?bid=3167167&fileBranchId=3175304#l562)

**Comment by lerks** [comment]
Hey, sorry for letting you wait so long to have a review (we're very short on manpower). Thanks for your contribution!

I like your idea of putting the validation at a low level. The way you went in this PR was what I had first in mind, but then I wondered if we could go any lower and in fact we can: we can have PostgreSQL enforce the constraint internally, so that the database is consistent even if accessed without SQLAlchemy. One way to do that is as follows:
```
codename = Column(
    Unicode,
    CheckConstraint("codename ~ '^[^a-zA-Z0-9_.-]+$'"),
    nullable=False)
```

Now, there's two more improvements I'd like to have. First, there are many fields whose alphabet should probably be restricted (filenames, usernames, task and contest codenames, and many more). It would be nice if we could have a decorator that put before the field achieves the same effect of putting that CheckConstraint (I'm not sure it's possible, but it's worth investigating).

Another thing I'm wondering about is what alphabet we want to allow. These codenames (of all sorts) appear in all kinds of places: file path components on disk, URL components, IDs of HTML tags, JSON dictionary keys, etc. I don't know exactly what constraints are enforced by each of these, but I'd rather err on the safe side. We should discuss this.

**Comment by lerks** [comment]
Also, I am not sure how the violation of a CheckConstraint shows up in AWS but it would be nice to fail gracefully, so something along the lines of https://github.com/myungwoo/cms/commit/0b6d94c may be desirable as well (in addition to what I said before, not instead of it).

**Comment by lerks** [comment]
I started working on this myself. Current progress is here: https://github.com/lerks/cms/tree/safe_codenames.

**Comment by wil93** [comment]
> we can have PostgreSQL enforce the constraint internally, so that the database is consistent even if accessed without SQLAlchemy

By the way, not really related to this PR but: this could also be done for `default` attributes by using `server_default` instead. I think it would make sense to do it.

**Comment by lerks** [comment]
Yes we could. I think originally the reason we chose not to do that was because we had the default password for Users be a randomly generated string (I don't know if this is still the case) and therefore for that we had to use the Python-enabled SQLAlchemy client-side defaults, and it would have been weird to mix the two.

**Comment by lerks** [comment]
Done in #738.

**Pull request closed without merging**